### ISSUE TITLE
[RFC] No floaters, at all (in prices and critical calculation)

### ISSUE THREAD

**Issue by Nek-** [Sylius/Sylius#13821]
I figured out that TaxRate is stored as a float. This leads developers to make calculation in floats (something like price / 100 * taxRate) which is not recommended. [The associated calculator is also highly dangerous](https://github.com/Sylius/Sylius/blob/66348ff2e9e2ea12faef489950898ef351a305a0/src/Sylius/Component/Taxation/Calculator/DefaultCalculator.php#L20-L27) and will give a lot of wrong results.

I suggest considering the migration to string for this.

The first action to take in that way would be to add getters returning string instead of float and deprecated float getters.

WDYT?

**Comment by mbabker** [comment]
As tax rates are usually a percentage, how do you store them without losing decimal precision?  Even if through the database schema they're stored in `VARCHAR` fields as strings and the PHP data model treats those values as strings, ultimately you're going to end up with some kind of arithmetic operation that uses a floating point number.

**Comment by Nek-** [comment]
Any calculation should be done using a math library:
- [bcmath](https://www.php.net/manual/fr/function.bcmul) : use a string (not a float) as input
- [brik/math](https://github.com/brick/math) accept string in construct `BigNumber::of($stringValue)`
- [gmp](https://www.php.net/manual/fr/function.gmp-mul.php) : accept strings or int (not floats)

So yes, you are going to do arithmetic, but it mostly requires string. Float always ends up with errors at some point in the life of the application. My suggestion is to ban it for anything that needs precision. (we agree that tax computation requires precision, right?)

**Comment by mbabker** [comment]
Having that helps.  I read this at first and thought you were saying just to change the data types (which seemed like change for the sake of it without the extra context), I didn't realize you were also suggesting internal changes as well.

**Comment by vvasiloi** [comment]
Here's a relevant issue reported via Slack ([full thread](https://sylius-devs.slack.com/archives/C3EGDG9LY/p1648827780599519)): 
![image](https://user-images.githubusercontent.com/7114562/161611854-734113d7-52a7-4b82-92b8-0031c6e7a120.png)
And a POC I made attempting to solve the issue within the current constraints: https://github.com/Sylius/Sylius/compare/master...vvasiloi:decimal-precision-tax

**Comment by lchrusciel** [comment]
The idea is neat but requires probably deep research. Some problems with it would be solved if decide to incorporate MoneyPHP. Keeping percentage in the DB as ints is ok, but won't change too much for me on its own. Some specialized math lib is probably even more important 

According to the issue @vvasiloi mentioned, for me, it could be solved with more precision on money counting - like with 4 digits under the hood. It is not a problem for Sylius as we are not divining money almost in any place. MoneyPHP would solve it as well, I assume - but that something I wouldn't count on in the nearest future

**Comment by Nek-** [comment]
TL;DR: string will work with everything, this is why string should be used here IMO.

> Keeping percentage in the DB as ints is ok

The suggestion here is string (btw MoneyPHP is one more string addict library) on purpose; the percentage needs to be as many decimals as we want. The suggestion of string is not innocent at all, it's what math libraries use in PHP for storage (as explained before).
